// @deck.gl/carto/dist/es5/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getDefaultCredentials", {
  enumerable: true,
  get: function get() {
    return _config.getDefaultCredentials;
  }
});
Object.defineProperty(exports, "setDefaultCredentials", {
  enumerable: true,
  get: function get() {
    return _config.setDefaultCredentials;
  }
});
Object.defineProperty(exports, "_getMapsVersion", {
  enumerable: true,
  get: function get() {
    return _config.getMapsVersion;
  }
});
Object.defineProperty(exports, "CartoSQLLayer", {
  enumerable: true,
  get: function get() {
    return _cartoSqlLayer["default"];
  }
});
Object.defineProperty(exports, "CartoBQTilerLayer", {
  enumerable: true,
  get: function get() {
    return _cartoBqtilerLayer["default"];
  }
});
Object.defineProperty(exports, "BASEMAP", {
  enumerable: true,
  get: function get() {
    return _basemap["default"];
  }
});
Object.defineProperty(exports, "colorBins", {
  enumerable: true,
  get: function get() {
    return _colorBinsStyle["default"];
  }
});
Object.defineProperty(exports, "colorCategories", {
  enumerable: true,
  get: function get() {
    return _colorCategoriesStyle["default"];
  }
});
Object.defineProperty(exports, "colorContinuous", {
  enumerable: true,
  get: function get() {
    return _colorContinuousStyle["default"];
  }
});
Object.defineProperty(exports, "_getTileJSON", {
  enumerable: true,
  get: function get() {
    return _mapsApiClient.getTileJSON;
  }
});

var _config = require("./config.js");

var _cartoSqlLayer = _interopRequireDefault(require("./layers/carto-sql-layer"));

var _cartoBqtilerLayer = _interopRequireDefault(require("./layers/carto-bqtiler-layer"));

var _basemap = _interopRequireDefault(require("./basemap"));

var _colorBinsStyle = _interopRequireDefault(require("./style/color-bins-style"));

var _colorCategoriesStyle = _interopRequireDefault(require("./style/color-categories-style"));

var _colorContinuousStyle = _interopRequireDefault(require("./style/color-continuous-style"));

var _mapsApiClient = require("./api/maps-api-client");
