// @deck.gl/carto/dist/es5/layers/carto-sql-layer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _cartoLayer = _interopRequireDefault(require("./carto-layer"));

var _config = require("../config");

var _mapsApiClient = require("../api/maps-api-client");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var defaultProps = {
  bufferSize: 16,
  tileExtent: 4096,
  uniqueIdProperty: 'cartodb_id'
};

var CartoSQLLayer = function (_CartoLayer) {
  (0, _inherits2["default"])(CartoSQLLayer, _CartoLayer);

  var _super = _createSuper(CartoSQLLayer);

  function CartoSQLLayer() {
    (0, _classCallCheck2["default"])(this, CartoSQLLayer);
    return _super.apply(this, arguments);
  }

  (0, _createClass2["default"])(CartoSQLLayer, [{
    key: "updateTileJSON",
    value: function () {
      var _updateTileJSON = (0, _asyncToGenerator2["default"])(_regenerator["default"].mark(function _callee() {
        var _this$props, data, bufferSize, tileExtent, credentials, version, isSQL, sql, mapConfig;

        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _this$props = this.props, data = _this$props.data, bufferSize = _this$props.bufferSize, tileExtent = _this$props.tileExtent, credentials = _this$props.credentials;
                version = (0, _config.getMapsVersion)(credentials);
                isSQL = data.search(' ') > -1;
                _context.t0 = version;
                _context.next = _context.t0 === 'v1' ? 6 : _context.t0 === 'v2' ? 11 : 14;
                break;

              case 6:
                sql = isSQL ? data : "SELECT * FROM ".concat(data);
                mapConfig = {
                  version: '1.3.1',
                  buffersize: {
                    mvt: bufferSize
                  },
                  layers: [{
                    type: 'mapnik',
                    options: {
                      sql: sql.trim(),
                      vector_extent: tileExtent
                    }
                  }]
                };
                _context.next = 10;
                return (0, _mapsApiClient.getTileJSON)({
                  mapConfig: mapConfig,
                  credentials: credentials
                });

              case 10:
                return _context.abrupt("return", _context.sent);

              case 11:
                _context.next = 13;
                return (0, _mapsApiClient.getTileJSON)({
                  connection: _mapsApiClient.CONNECTIONS.CARTO,
                  source: data,
                  type: isSQL ? _mapsApiClient.MAP_TYPES.SQL : _mapsApiClient.MAP_TYPES.TABLE,
                  credentials: credentials
                });

              case 13:
                return _context.abrupt("return", _context.sent);

              case 14:
                throw new Error("Cannot build MapConfig for unmatching version ".concat(version));

              case 15:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function updateTileJSON() {
        return _updateTileJSON.apply(this, arguments);
      }

      return updateTileJSON;
    }()
  }]);
  return CartoSQLLayer;
}(_cartoLayer["default"]);

exports["default"] = CartoSQLLayer;
CartoSQLLayer.layerName = 'CartoSQLLayer';
CartoSQLLayer.defaultProps = defaultProps;
