// @deck.gl/carto/dist/esm/layers/carto-bqtiler-layer.js
import _regeneratorRuntime from "@babel/runtime/regenerator";
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import CartoLayer from './carto-layer';
import { getTileJSON, CONNECTIONS, MAP_TYPES } from '../api/maps-api-client';

var CartoBQTilerLayer = function (_CartoLayer) {
  _inherits(CartoBQTilerLayer, _CartoLayer);

  var _super = _createSuper(CartoBQTilerLayer);

  function CartoBQTilerLayer() {
    _classCallCheck(this, CartoBQTilerLayer);

    return _super.apply(this, arguments);
  }

  _createClass(CartoBQTilerLayer, [{
    key: "updateTileJSON",
    value: function () {
      var _updateTileJSON = _asyncToGenerator(_regeneratorRuntime.mark(function _callee() {
        var _this$props, credentials, data;

        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _this$props = this.props, credentials = _this$props.credentials, data = _this$props.data;
                _context.next = 3;
                return getTileJSON({
                  connection: CONNECTIONS.BIGQUERY,
                  type: MAP_TYPES.TILESET,
                  source: data,
                  credentials: credentials
                });

              case 3:
                return _context.abrupt("return", _context.sent);

              case 4:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function updateTileJSON() {
        return _updateTileJSON.apply(this, arguments);
      }

      return updateTileJSON;
    }()
  }]);

  return CartoBQTilerLayer;
}(CartoLayer);

export { CartoBQTilerLayer as default };
CartoBQTilerLayer.layerName = 'CartoBQTilerLayer';
